lib/src/klio/utils/_thread_limiter.py

```diff
@@ -0,0 +1,157 @@
+# Copyright 2020 Spotify AB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import enum
+import logging
+import multiprocessing
+import threading
+
+
+class _DummySemaphore(object):
+    """Mock semaphore for API parity when no limit is set in ThreadLimiter."""
+
+    def __init__(self, *args, **kwargs):
+        self._value = "NONE"
+
+    def acquire(self):
+        pass
+
+    def release(self):
+        pass
+
+
+class ThreadLimit(enum.Enum):
+    """:class:`enum.Enum` of constants for :class:`ThreadLimiter`.
+
+    Example usage:
+
+    .. code-block:: python
+
+        from klio import utils
+
+        thread_limiter = utils.ThreadLimiter(
+            max_thread_limit=utils.ThreadLimit.DEFAULT
+        )
+    """
+
+    NONE = 0
+    """Do not limit the number of threads used."""
+    DEFAULT = 1
+    """Default thread limit (CPU count of worker machine via
+    :func:`multiprocessing.cpu_count`).
+    """
+
+
+class ThreadLimiter(object):
+    """A context manager to limit active threads for a block of code.
+
+    Threads aren't limited `directly`; threads are limited indirectly
+    by maintaining a semaphore (an atomic counter).
+
+    Example usage:
+
+    .. code-block:: python
+
+        # No arg to use the default limit (CPU count of host/worker):
+        thread_limiter = ThreadLimiter()
+
+        # Then use it as a context manager:
+        with thread_limiter:
+            ...
+
+        # Or explicitly call acquire and release:
+        thread_limiter.acquire()
+        ...
+        thread_limiter.release()
```

Comment:
If we're not explicitly calling out the fact that we're using a `BoundedSemaphore` under the hood, it could be good to say that when the `thread_limiter` is released too many times, a `ValueError` will be raised.